README.md
[HistoryDB](http://doc.reverbrain.com/historydb:historydb)
=========

[History DB](http://doc.reverbrain.com/historydb:historydb) is a trully scalable (hundreds of millions updates per day) distributed archive system
with per user and per day activity statistics.

[History DB](http://doc.reverbrain.com/historydb:historydb) uses [elliptics](http://reverbrain.com/elliptics) as storage backend.
It supports 2 record types: user logs (updated via append or write) and activity.
Each user logs file hosts user logs for certain user and for specified period (by default this equals ot 1 day).
Activity is secondary index which hosts information about
who has activity during specified period (by default this equals to 1 day).

User log primary key is `username (string prefix) + '.' + timestamp (of the day)`.

Activity primary key is `timestamp (of the day) + '.' + chunk number`.

All daily activity is devided to chunks.
One can specify own activity prefix if needed.

User log is a blob which can be either appended or rewritten.


[User guide](http://doc.reverbrain.com/historydb:cplusplus_api)
===========

Interface of the [History DB](http://doc.reverbrain.com/historydb:historydb) presented in `provider.h` file.

	provider::provider() - contructor
	
	provider::set_session_parameters() - sets parameters for all sessions.
		It includes vector of elliptics groups (replicas) in which HistoryDB stores data and
		minimum number of succeded writes.

	provider::add_log - appends data to user log

	provider::add_activity - updates user activity
	
	provider::add_log_with_activity - appends data to user log and updates user activity

	provider::get_user_logs() - gets user logs.

	provider::get_active_user() - gets active user for specified day.

	provider::for_user_logs() - iterates over user's logs in specified time period.
	
	provider::for_active_user() - iterates over activity logs in specified time period.

One can grab user logs for specified for specified period of time as well as list of all users,
who were active (had at least one log update) during requested period of time.

[Tutorial](http://doc.reverbrain.com/historydb:tutorial)
=========
See [extended tutorial](http://doc.reverbrain.com/historydb:tutorial)

Firstly, complete [elliptics tutorial](http://doc.reverbrain.com/elliptics:server-tutorial).
It is neccessary to start work with elliptics.

Download source tree

	git clone http://github.com/reverbrain/historydb.git

Building library

	cd historydb
	debuild
	sudo dpkg -i ../historydb_*.deb

Now you can start to use HistoryDB library:
Include `historydb/provider.h`.
Create `provider` instance.
Use provider instance to write/read user logs, update activity, gets active user statistics, repartition shards etc.

[HTTP interface](http://doc.reverbrain.com/historydb:http_api)
=========
[HistoryDB](http://doc.reverbrain.com/historydb:historydb) has follow HTTP interface implemented via fastcgi-daemon2:

	"/add_log" POST - adds record to user logs.
		Parameters:
			user - name of the user
			data - data of the log record
			time or key. If both: key and time are specified - key will be used
				time - timestamp of record
				key - custom key of record

	"/add_activity" POST - marks user as active in the day.
		Parameters:
			user - name of the user
			time or key. If both: key and time are specified - key will be used
				time - timestamp of activity statistics
				key - custom key of activity statistics
				
	"/add_log_with_activity" POST - Appends log to user logs and updates user activity.
		Parameters:
			user - name of user
			data - data of log record
			time or key. If both: key and time are specified - key will be used
				time - timestamp of record
				key - custom key of record
	
	"/get_active_users" GET - returns users who was active in the day.
		Parameters:
			time or key. If both: key and time are specified - key will be used
				time - timestamp of activity statistics
				key - custom key of activity statistics
	
	"/get_user_logs" GET - returns logs of user.
		Parameters:
			user - name of the user
			begin_time and end_time - time period for logs
			
	"/" POST&GET - has no parameters. If all is ok - returns HTTP 200. May be used for checking service.

[Fastcgi-daemon2 config file](http://doc.reverbrain.com/historydb:http_configure)
=========

[HistoryDB](http://doc.reverbrain.com/historydb:historydb) component element should have follow children:

<pre>&lt;log_file&gt;/path/to/log_file&lt;/log_file&gt; - path to elliptics client logs

&lt;log_level&gt;LOG_LEVEL&lt;/log_level&gt; - valid values = { DATA, ERROR, INFO, NOTICE, DEBUG }

&lt;elliptics&gt; - one &lt;elliptics&gt; for each elliptics node
	&lt;addr&gt;address&lt;/addr&gt; - address of elliptics node
	&lt;port&gt;port&lt;/port&gt; - listening port on elliptics node
	&lt;family&gt;family&lt;/family&gt; - protocol family
&lt;/elliptics&gt;

&lt;group&gt;group_number&lt;/group&gt; - group number with which historydb will works. One <group> for each elliptics group.

&lt;min_writes&gt;number&lt;/min_writes&gt; - minimum number of succeded writes in groups. For example, if historydb tries to write in 5 groups and min_writes is 3
the attemp will be failed if write will be succeded in less then 3 groups.
</pre>

[HistoryDB Tool for aggregacting logs](http://doc.reverbrain.com/historydb:tools)
=========
The tool goes through activity statisitics for specified days, aggregates logs for each active users and saves it in specified new subkeys.

	Usage: hdb_tool.py KEYS NEW_KEY [options]

	Options:
		-h, --help            show this help message and exit
		-b BATCH_SIZE, --batch-size=BATCH_SIZE
			Number of keys in read_bulk/write_bulk batch [default:1024]
		-d, --debug           Enable debug output [default: False]
		-r ELLIPTICS_REMOTE, --remote=ELLIPTICS_REMOTE
			Elliptics node address [default: none]
		-g ELLIPTICS_GROUPS, --groups=ELLIPTICS_GROUPS
			Comma separated list of groups [default: all]
		-l FILE, --log=FILE   Output log messages from library to file [default:hdb_tool.log]
		-L ELLIPTICS_LOG_LEVEL, --log-level=ELLIPTICS_LOG_LEVEL
			Elliptics client verbosity [default: 1]
		-u USERS, --user=USERS
			User whose logs should be aggregated


debian/changelog
historydb (0.3.0.0) unstable; urgency=low

  * Moved to new version of thevoid: 0.6.1.3 

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Thu, 06 Mar 2014 20:29:53 +0400

historydb (0.2.6.0) unstable; urgency=low

  * Used ioremap::elliptics::data_pointer for logs data.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Wed, 30 Oct 2013 02:25:25 +0400

historydb (0.2.5.2) unstable; urgency=low

  * Added ability to use void*/size_t or std::string for logs data.
  * Added add_log_with_activity to fastcgi.
  * Increased version of thevoid, swarm and elliptics in deps. 

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 28 Oct 2013 22:38:54 +0400

historydb (0.2.5.1) unstable; urgency=low

  * Updated thevoid and elliptics version.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 14 Oct 2013 15:47:47 +0400

historydb (0.2.5.0) unstable; urgency=low

  * Compatibility with elliptics 2.24.14.15

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 30 Sep 2013 17:16:28 +0400

historydb (0.2.4.2) unstable; urgency=low

  * Fixed error handling while async add_log_with_activity.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Thu, 08 Aug 2013 16:33:46 +0400

historydb (0.2.4.1) unstable; urgency=low

  * Revert "Get rid of cmake-generated files"

 -- Evgeniy Polyakov <zbr@ioremap.net>  Tue, 06 Aug 2013 18:24:19 +0400

historydb (0.2.4.0) unstable; urgency=low

  * Enabled cache for adding logs and activity.
  * Added logs for async operation.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Fri, 02 Aug 2013 13:34:37 +0400

historydb (0.2.3.0) unstable; urgency=low

  * Added hand for appending log and updating activity
  * Updated version of elliptics.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Thu, 11 Jul 2013 12:45:58 +0400

historydb (0.2.2.0) unstable; urgency=low

  * Removed sharding activity on historydb side. Updated elliptics depency

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 08 Jul 2013 14:34:57 +0400

historydb (0.2.1.2) unstable; urgency=low

  * Removed cerr from HistoryDB-TheVoid initialization

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Fri, 28 Jun 2013 20:50:49 +0400

historydb (0.2.1.1) unstable; urgency=low

  * New version of swarm: 0.5.1

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Fri, 28 Jun 2013 20:39:38 +0400

historydb (0.2.1.0) unstable; urgency=low

  * Increased timeout for elliptics operation.
  * Implemented HistoryDB-TheVoid web-server.
  * Updated HistoryDB HTTP and C++ API.
  * Added pre-version of combine tool.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Fri, 28 Jun 2013 15:25:29 +0400

historydb (0.2.0.4) unstable; urgency=low

  * Fixed missing Content-Length in HTTP response header.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Fri, 21 Jun 2013 00:06:52 +0400

historydb (0.2.0.3) unstable; urgency=low

  * Added extra exception check while adding data to user log.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 10 Jun 2013 21:04:54 +0400

historydb (0.2.0.2) unstable; urgency=low

  * Fixed http api. Changed missed parameters http code.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Wed, 05 Jun 2013 20:19:10 +0400

historydb (0.2.0.1) unstable; urgency=low

  * Fixed activity statistics methods

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Tue, 04 Jun 2013 21:07:45 +0400

historydb (0.2.0.0) unstable; urgency=low

  * New API implementation. Removed inheritance and virtual functions.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 03 Jun 2013 20:04:47 +0400

historydb (0.1.1.1) unstable; urgency=low

  * Fixed spec file

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Tue, 07 May 2013 12:22:15 +0400

historydb (0.1.1.0) unstable; urgency=low

  * Implemented fastcgi proxy library and include it in debian package.
  * Chose json as output format for fastcgi proxy.
  * Replaced using write_cas'es to using indexes.
  * Improved configuring HistoryDB via fastcgi config file and c++ interface.
  * Extended logging HistoryDB.
  * Fixed build on rhel 5/6.
  * Added separated methods for adding user log and incrementing user activity.


 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 06 May 2013 16:43:36 +0400

historydb (0.1.0.0) unstable; urgency=low

  * History DB is a trully scalable (hundreds of millions updates per day) distributed archive system
  * with per user and per day activity statistics.

 -- Kirill Smorodinnikov <shaitan@yandex-team.ru>  Mon, 1 Apr 2010 18:32:20 +0400


tools/hdb_tool.py
#!/usr/bin/env python

# =============================================================================
#  Copyright 2013+ Kirill Smorodinnikov <shaitkir@gmail.com>
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# =============================================================================


import sys
import logging
import logging.handlers
from itertools import groupby

import elliptics
from elliptics_recovery.route import Address

log = logging.getLogger()
log.setLevel(logging.DEBUG)
formatter = logging.Formatter(fmt='%(asctime)-15s %(processName)s %(levelname)s %(message)s',
                              datefmt='%d %b %y %H:%M:%S')

ch = logging.StreamHandler(sys.stderr)
ch.setFormatter(formatter)
ch.setLevel(logging.INFO)
log.addHandler(ch)


def combine_logs(users, keys, new_key, batch_size, node, groups):
    log.debug("Creating session for reading and appending logs")
    log_session = elliptics.Session(node)
    log_session.ioflags = elliptics.io_flags.append
    log_session.cflags = elliptics.command_flags.nolock
    log_session.groups = groups

    log.debug("Creating session for reading and updating activity")
    activity_session = elliptics.Session(node)
    activity_session.cflags = elliptics.command_flags.nolock
    activity_session.groups = groups

    for key in keys:
        process_key(users=users,
                    key=key,
                    new_key=new_key,
                    batch_size=batch_size,
                    log_session=log_session,
                    activity_session=activity_session)


def process_key(users, key, new_key, batch_size, log_session, activity_session):
    log.debug("Processing key: {0}".format(key))
    try:
        results = activity_session.find_any_indexes([key]).get()
        log.debug("Found {0} users active at {1}".format(len(results), key))
        for batch_id, batch in groupby(enumerate((r.indexes[0].data for r in results)), key=lambda x: x[0] / batch_size):
            if users and len(users):
                batch = users.intersection([u for _, u in batch])
            else:
                batch = [u for _, u in batch]
            process_users(batch, key, new_key, log_session, activity_session)
    except Exception as e:
        log.error("Coudn't process key: {0} error: {1}".format(key, e))
    except:
        log.error("Coudn't process key: {0}".format(key))


def process_users(users, key, new_key, log_session, activity_session):
    users_len = len(users)
    log.debug("Processing users: {0} for key: {1}".format(users_len, key))

    if not users or len(users) == 0:
        log.debug("No users specified, skipping")
        return

    log.debug("Async updating indexes for {0} users for index: {1}".format(users_len, new_key))
    async_indexes = []
    for u in users:
        async_indexes.append(activity_session.update_indexes_internal(elliptics.Id(u), [new_key], [u]))

    log.debug("Async reading logs for {0} users".format(users_len))
    async_read = log_session.bulk_read_async([elliptics.Id(u + '.' + key) for u in users])

    log.debug("Async writing read logs")
    async_writes = []
    it = iter(async_read)
    failed = 0
    while True:
        try:
            r = next(it)
            async_writes.append((log_session.write_data_async((r.id, elliptics.Time(2**64-1, 2**64-1), 0), r.data), len(r.data)))
        except StopIteration:
            break
        except Exception as e:
            failed += 1
            log.debug("Write failed: {0}".format(e))

    writed_bytes = 0
    for r, size in async_writes:
        r.wait()
        if r.successful():
            writed_bytes += size
        else:
            failed += 1

    log.info("Aggregated bytes: {0}".format(writed_bytes))
    log.info("Successed: {0}".format(users_len))
    log.info("Failures: {0}".format(failed))

    successes, failures = (0, 0)
    for a in async_indexes:
        a.wait()
        if a.successful():
            successes += 1
        else:
            failures += 1
    log.info("Activtiy updates successes: {0}".format(users_len))
    log.info("Activtiy updates failures: {0}".format(failed))


if __name__ == '__main__':
    from optparse import OptionParser

    parser = OptionParser()
    parser.usage = "%prog KEYS NEW_KEY [options]"
    parser.description = __doc__
    #parser.add_option("-n", "--nprocess", action="store", dest="nprocess", default="1",
    #                  help="Number of subprocesses [default: %default]")
    parser.add_option("-b", "--batch-size", action="store", dest="batch_size", default="1024",
                      help="Number of keys in read_bulk/write_bulk batch [default: %default]")
    parser.add_option("-d", "--debug", action="store_true", dest="debug", default=False,
                      help="Enable debug output [default: %default]")
    parser.add_option("-r", "--remote", action="store", dest="elliptics_remote", default=None,
                      help="Elliptics node address [default: %default]")
    parser.add_option("-g", "--groups", action="store", dest="elliptics_groups", default=None,
                      help="Comma separated list of groups [default: all]")
    parser.add_option("-l", "--log", dest="elliptics_log", default='hdb_tool.log', metavar="FILE",
                      help="Output log messages from library to file [default: %default]")
    parser.add_option("-L", "--log-level", action="store", dest="elliptics_log_level", default="1",
                      help="Elliptics client verbosity [default: %default]")
    parser.add_option("-u", "--user", action="append", dest="users", default=[],
                      help="User whose logs should be aggregated")

    (options, args) = parser.parse_args()

    ch.setLevel(logging.WARNING)
    if options.debug:
        ch.setLevel(logging.DEBUG)

    log.info('Initializing')

    try:
        log_file = options.elliptics_log
        log_level = int(options.elliptics_log_level)
    except Exception as e:
        raise ValueError("Can't parse log_level: '{0}': {1}".format(options.elliptics_log_level, e))
    log.info("Using elliptics client log level: {0}".format(log_level))

    try:
        batch_size = int(options.batch_size)
        if batch_size <= 0:
            raise ValueError("Batch size should be positive: {0}".format(batch_size))
    except Exception as e:
        raise ValueError("Can't parse batchsize: '{0}': {1}".format(options.batch_size, e))
    log.info("Using batch_size: {0}".format(batch_size))

    if not options.elliptics_remote:
        raise ValueError("Elliptics address should be given (-r option).")
    try:
        address = Address.from_host_port_family(options.elliptics_remote)
    except Exception as e:
        raise ValueError("Can't parse host:port:family: '{0}': {1}".format(options.elliptics_remote, e))
    log.info("Using host:port:family: {0}".format(address))

    if not options.elliptics_groups:
        raise ValueError("Elliptics groups should be given (-g option).")
    try:
        groups = map(int, options.elliptics_groups.split(','))
    except Exception as e:
        raise ValueError("Can't parse grouplist: '{0}': {1}".format(options.elliptics_groups, e))
    log.info("Using group list: {0}".format(groups))

    try:
        users = set(options.users)
    except Exception as e:
        raise ValueError("Can't parse keys: '{0}': {1}".format(options.users, e))
    log.info("Aggregating keys for users: 0".format(users))

    try:
        keys = args[0].split(':')
    except Exception as e:
        raise ValueError("Can't parse keys: '{0}': {1}".format(args[0], e))
    log.info("Aggregating keys: 0".format(keys))

    try:
        new_key = args[1]
    except Exception as e:
        raise ValueError("Can't parse new_key: '{0}': {1}".format(args[1], e))
    log.info("Aggregating keys in key: 0".format(new_key))

    log.info("Setting up elliptics client")

    log.debug("Creating logger")
    elog = elliptics.Logger(log_file, int(log_level))

    log.info("Creating node using: {0}".format(address))
    cfg = elliptics.Config()
    cfg.config.wait_timeout = 60
    cfg.config.check_timeout = 60
    cfg.config.io_thread_num = 16
    cfg.config.nonblocking_io_thread_num = 16
    cfg.config.net_thread_num = 16
    node = elliptics.Node(elog, cfg)
    node.add_remote(addr=address.host, port=address.port, family=address.family)
    log.info("Created node: {0}".format(node))

    log.info("Keys which will be aggregated: {0}".format(keys))
    log.info("New aggregated key: {0}".format(new_key))

    combine_logs(users, keys, new_key, batch_size, node, groups)
